### PR TITLE
Fix resteasy-reactive-client hostname verification default

### PR DESCRIPTION
**Pull request by polarctos**
Since introduction of the setting [`verifyHost` in `2.15.0.Final`](https://github.com/quarkusio/quarkus/issues/27901) the hostname verification was disabled by default for the `resteasy-reactive-client`, as the [default value for `boolean` in Java](https://docs.oracle.com/javase/specs/jls/se17/html/jls-4.html#jls-4.12.5) (primitive) is `false`.
https://github.com/gsmet/quarkus/commit/941d3a6f7a1857d01a1a2b0eada67e3628895c0c#diff-28d8221e6b920ee90f85249d3c250b932c207d6b2099fc9377984d7f550fe28d

A better configuration name value would have been a negated name like `disableHostnameVerification`, but renaming it is no longer possible as it would break compatibility.

This disabled default currently makes the reactive client vulnerable to MITM attacks. 
E.g. the following [rest client builder from the documentation](https://quarkus.io/guides/rest-client-reactive#programmatic-client-creation-with-restclientbuilder) with `quarkus-rest-client-reactive-jackson` is vulnerable, as it does not fail during the SSL handshake:
```java
RestClientBuilder.newBuilder()
    .baseUri(URI.create("https://wrong.host.badssl.com"))
    .build(WrongHostClient.class);
```
The expected exception has the cause `javax.net.ssl.SSLHandshakeException`. The test case `ExternalWrongHostTestCase.java` currently only tests the disabling of hostname verification but not for a default secure behaviour.

In the meantime setting the config explicitly is a workaround e.g. with `quarkus.rest-client.verify-host=true`.
This enables hostname verification again also for the reactive client. But the default of Quarkus should instead be secure!

This change now adds a proper default both in the configuration and assigns safe initial values for the fields in the reactive client builder implementation.

Both custom `HostnameVerifier` and `SSLContext` are currently not supported by resteasy-reactive-client. 
This is documented as [known limitations](https://quarkus.io/guides/rest-client-reactive#known-limitations). 

This change also aligns the implementation for both to the already present state for `SSLContext`. 
It makes the builder fail when `HostnameVerifier` is tried to be configured. The benefit is that it fails as early as
currently possible and not only when the custom hostname verification would be needed later on.

The general need for the configuration option of a custom hostname verifier still stays relevant, as it was also requested in https://github.com/quarkusio/quarkus/issues/27901. Currently just having the possibility to completely disable hostname verification is preventing the chance to supply a suitable customised hostname verifier that implements custom verification rules.

### CODE REVIEW

File: extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java [hunk 1]
```diff
@@ -227,11 +227,12 @@ public class RestClientsConfig {
     public Optional<QueryParamStyle> queryParamStyle;
 
     /**
-     * Set whether hostname verification is enabled.
+     * Set whether hostname verification is enabled. Default is enabled.
+     * This setting should not be used in production as it makes the client vulnerable to MITM attacks.
```
Comment (gastaldi) [inline extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java:231]:
```suggestion
     * This setting should not be disabled in production as it makes the client vulnerable to MITM attacks.
```

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java [hunk 2]
```diff
@@ -114,8 +114,8 @@ public ClientBuilder trustStore(KeyStore trustStore, char[] password) {
 
     @Override
     public ClientBuilder hostnameVerifier(HostnameVerifier verifier) {
-        this.hostnameVerifier = verifier;
-        return this;
+        // TODO
+        throw new RuntimeException("Specifying HostnameVerifier is not supported at the moment");
```
Comment (gastaldi) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java:118]: This doesn't look right to me

Comment (polarctos) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java:118]: I first thought the same, but this is exactly was the existing code does for the same unsupported feature of custom `SSLContext` for the reactive client.

https://github.com/quarkusio/quarkus/blob/3.0.1.Final/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java#L89


This is existing code from the same class:
````java
    @Override
    public ClientBuilder sslContext(SSLContext sslContext) {
        // TODO
        throw new RuntimeException("Specifying SSLContext is not supported at the moment");
    }
````

From my perspective failing early here is actually better than the current code, that makes it look like the hostname verifier is applied, but then never actually used by the reactive client.

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java [hunk 3]
```diff
@@ -66,8 +66,8 @@ public class ClientBuilderImpl extends ClientBuilder {
     private String nonProxyHosts;
 
     private boolean followRedirects;
-    private boolean trustAll;
-    private boolean verifyHost;
+    private boolean trustAll = false;
```
Comment (gastaldi) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java:69]: This is false by default, no need to change it

Comment (polarctos) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java]: Fixed in https://github.com/quarkusio/quarkus/pull/32932/commits/7babb3221397a336d33cf7c155642230878d97fc

**Review by Sgitario — APPROVED** [review]
lgtm

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java (hunk 2 above)
Comment (gastaldi) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java:118]: How about throwing an `UnsupportedOperationException` instead?

**Comment by geoand** [comment]
The test failures are definitely related

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java (hunk 2 above)
Comment (polarctos) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java]: That fits better. Thanks. Switched both for `UnsupportedOperationException` to keep it consistent.

Fixed in https://github.com/quarkusio/quarkus/pull/32932/commits/7f7d717b54cbaf8f03f00c33f1b3aa987e54cbaf

**Comment by polarctos** [comment]
> The test failures are definitely related

@geoand 

Sorry for that. Did not expect there to be tests for an unsupported feature.
I now removed the unsupported config tests for setting a `hostnameVerifier`.

On the `rest-client-reactive` config a custom hostname verifier was also before not supported, but did not throw an exception. Adapted the tests accordingly as the builder fails early now throwing `UnsupportedOperationException`.

The general discussion if failing here with an exception makes sense is discussed here: https://github.com/quarkusio/quarkus/pull/32932#discussion_r1178488500

In short: It should be the same behaviour for both unsupported features of the `rest-client-reactive` of `HostnameVerifier` and `SSLContext`. 
Long term it would actually make sense to try to implement them, as they are part of the [MP RestClientBuilder API](https://github.com/eclipse/microprofile-rest-client/blob/3.0.1/api/src/main/java/org/eclipse/microprofile/rest/client/RestClientBuilder.java#L216).

**Comment by polarctos** [comment]
Should a CVE be opened for this for the affected versions so that affected users are notified by scanners?

The actual weakness would be local to default `false` value of `verifyHost` 
in `org.jboss.resteasy.reactive.client.impl.ClientBuilderImpl` 
from `io.quarkus.resteasy.reactive:resteasy-reactive-client`.

From first sight the affected versions include: `>= 2.15.0.Final <= 3.0.1.Final`

At least for other products CVEs were sometimes opened fur such trivial things, 
e.g. where in the defaults it is running `curl -k`.

Applicable CWEs:
https://cwe.mitre.org/data/definitions/295.html
https://cwe.mitre.org/data/definitions/453.html

**Comment by geoand** [comment]
As I have 0️⃣ time for at least the next week, I'll leave this up to @Sgitario and @sberyozkin

**Comment by Sgitario** [comment]
> Should a CVE be opened for this for the affected versions so that affected users are notified by scanners?
> 
> The actual weakness would be local to default `false` value of `verifyHost` in `org.jboss.resteasy.reactive.client.impl.ClientBuilderImpl` from `io.quarkus.resteasy.reactive:resteasy-reactive-client`.
> 
> From first sight the affected versions include: `>= 2.15.0.Final <= 3.0.1.Final`
> 
> At least for other products CVEs were sometimes opened fur such trivial things, e.g. where in the defaults it is running `curl -k`.
> 
> Applicable CWEs: https://cwe.mitre.org/data/definitions/297.html / https://cwe.mitre.org/data/definitions/295.html https://cwe.mitre.org/data/definitions/453.html

This sounds reasonable to me +1

**Comment by polarctos** [comment]
I added a test now to assert that certificate host name verification is now enabled by default and working.

Build is also green now, based on the GitHub actions in the forked repository.

**Comment by geoand** [comment]
cc @Sgitario

**Review by Sgitario — COMMENTED** [review]
The changes look good to me.
Though, there are still some custom hostname verifier classes that should be deleted now:
- MyHostnameVerifier1 and MyHostnameVerifier2 in RestClientCDIDelegateBuilderTest
- MyHostnameVerifier in HelloClientWithBaseUri

Also, squash all the commits, so we can proceed with merging these changes.

Many thanks!

**Comment by polarctos** [comment]
> Though, there are still some custom hostname verifier classes that should be deleted now:
> 
> * MyHostnameVerifier1 and MyHostnameVerifier2 in RestClientCDIDelegateBuilderTest
> * MyHostnameVerifier in HelloClientWithBaseUri
> 
> Also, squash all the commits, so we can proceed with merging these changes.

@Sgitario I removed the unused hostname verifier classes from the tests now.

I squashed it down to two commits, but these two are actually independent of each other:
* Fail on attempt to set custom hostname verifier
* Enabling hostname verification by default

From my perspective, only the second commit for enabling the hostname verification is necessary as a security fix backport to other branches, as it fixes the current vulnerability.

**Comment by polarctos** [comment]
I think the pull request is now ready to be merged.

**Comment by Sgitario** [comment]
@gastaldi would you mind a final review of this one and merging it if you're ok with it? Thanks!

**Review by gastaldi — APPROVED** [review]

**Comment by sberyozkin** [comment]
Hi, should this PR check if `quarkus.tls.trust-all` is configured and if yes, prefer `quarkus.tls.trust-all` ? This is how OIDC client code is configured...

**Comment by sberyozkin** [comment]
I suppose it can be handled in a follow up PR. @polarctos would you be OK with opening a new issue related to checking `quarkus.tls.trust-all` ? Thanks

**Comment by Sgitario** [comment]
> Hi, should this PR check if `quarkus.tls.trust-all` is configured and if yes, prefer `quarkus.tls.trust-all` ? This is how OIDC client code is configured, so that if for example, OIDC protected endpoint is using OIDC reactive token propagation which relies on Resteasy Reactive client, the hostname verification can be configured with a single property, as opposed to configuring it twice

if 'verify-host` is true, but `trust-all` is true, then the `verify-host` will be turned off: https://github.com/quarkusio/quarkus/blob/5bca78de10fb7a6d6ddeecda86e31180ca8ffff6/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java#L213

So, I think it's fine how it is now.

**Review by sberyozkin — APPROVED** [review]

**Pull request merged**